File: deal-mgmt/app/schemas/document_version.py
"""문서 버전 관리 스키마 — DocumentMaster + DocumentRevision."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

# ── DocumentMaster ──────────────────────────────────────


class DocumentMasterCreate(BaseModel):
    doc_type: str = Field(..., description="DocumentType enum 값")
    doc_name: str = Field(..., min_length=1, max_length=500)
    description: str | None = None
    contract_id: uuid.UUID | None = None
    nda_id: uuid.UUID | None = None


class DocumentMasterUpdate(BaseModel):
    doc_name: str | None = Field(None, min_length=1, max_length=500)
    description: str | None = None
    is_archived: bool | None = None


class DocumentMasterOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    transaction_id: uuid.UUID
    doc_type: str
    doc_name: str
    description: str | None = None
    current_revision_number: int
    is_archived: bool
    contract_id: uuid.UUID | None = None
    nda_id: uuid.UUID | None = None
    created_by_email: str | None = None
    created_at: datetime
    updated_at: datetime


class DocumentMasterListResponse(BaseModel):
    items: list[DocumentMasterOut]
    total: int


# ── DocumentRevision ────────────────────────────────────


class DocumentRevisionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    revision_number: int
    sha256_hash: str
    file_name: str
    file_size_bytes: int
    mime_type: str | None = None
    uploaded_by_email: str | None = None
    upload_source: str
    changes_summary: str | None = None
    prev_revision_id: uuid.UUID | None = None
    is_current: bool
    is_deleted: bool
    source_entity_type: str | None = None
    source_entity_id: str | None = None
    created_at: datetime
    updated_at: datetime


class RevisionListResponse(BaseModel):
    items: list[DocumentRevisionOut]
    total: int

File: amic-platform/src/modules/ma/types/document_version.ts
// 문서 버전 관리 타입 — DocumentMaster + DocumentRevision

export type DocumentType =
  | "CONTRACT_SPA"
  | "CONTRACT_AMENDMENT"
  | "CONTRACT_SIDE_LETTER"
  | "CONTRACT_SHA"
  | "CONTRACT_ESCROW"
  | "CONTRACT_BTA"
  | "CONTRACT_SSA"
  | "CONTRACT_OTHER"
  | "NDA"
  | "RFI_EXCEL"
  | "MEETING_MINUTES"
  | "DD_REPORT"
  | "OTHER";

export type UploadSource =
  | "MANUAL"
  | "CONTRACT_MARKUP"
  | "NDA_MARKUP"
  | "RFI_IMPORT"
  | "SYSTEM";

export interface DocumentMaster {
  id: string;
  transaction_id: string;
  doc_type: DocumentType;
  doc_name: string;
  description: string | null;
  current_revision_number: number;
  is_archived: boolean;
  contract_id: string | null;
  nda_id: string | null;
  created_by_email: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentMasterListResponse {
  items: DocumentMaster[];
  total: number;
}

export interface DocumentRevision {
  id: string;
  document_id: string;
  revision_number: number;
  sha256_hash: string;
  file_name: string;
  file_size_bytes: number;
  mime_type: string | null;
  uploaded_by_email: string | null;
  upload_source: UploadSource;
  changes_summary: string | null;
  prev_revision_id: string | null;
  is_current: boolean;
  is_deleted: boolean;
  source_entity_type: string | null;
  source_entity_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface RevisionListResponse {
  items: DocumentRevision[];
  total: number;
}

export interface DocumentMasterCreatePayload {
  doc_type: DocumentType;
  doc_name: string;
  description?: string | null;
  contract_id?: string | null;
  nda_id?: string | null;
}
